hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMRecoverLeaseRequest.java

```diff
@@ -0,0 +1,264 @@
+            omResponse.setCmdType(RecoverLease);
+            omClientResponse = new OMRecoverLeaseResponse(
+                createErrorOMResponse(omResponse, ex), getBucketLayout(), null);
+        } finally {
+            if (omClientResponse != null) {
+                omClientResponse.setFlushFuture(
+                        ozoneManagerDoubleBufferHelper.add(omClientResponse,
+                                transactionLogIndex));
+            }
+            if (acquiredLock) {
+                omMetadataManager.getLock().releaseWriteLock(BUCKET_LOCK, volumeName,
+                        bucketName);
+            }
+        }
+
+        // Audit Log outside the lock
+        auditLog(ozoneManager.getAuditLogger(), buildAuditMessage(
+                OMAction.RECOVER_LEASE, auditMap, exception,
+                getOmRequest().getUserInfo()));
+
+        return omClientResponse;
+    }
+
+    private String doWork(OzoneManager ozoneManager,
+        RecoverLeaseRequest recoverLeaseRequest, long transactionLogIndex)
+        throws IOException {
+
+        //String dbOzoneKey =
+        //    omMetadataManager.getOzoneKey(volumeName, bucketName,
+        //        keyName);
+        final long volumeId = omMetadataManager.getVolumeId(volumeName);
+        final long bucketId = omMetadataManager.getBucketId(
+            volumeName, bucketName);
+        Iterator<Path> pathComponents = Paths.get(keyName).iterator();
+        long parentID = OMFileRequest.getParentID(volumeId, bucketId,
+            pathComponents, keyName, omMetadataManager,
+            "Cannot recover file : " + keyName
+                + " as parent directory doesn't exist");
+        String fileName = OzoneFSUtils.getFileName(keyName);
+        String dbFileKey = omMetadataManager.getOzonePathKey(volumeId, bucketId,
+            parentID, fileName);
+
+        OmKeyInfo keyInfo = getKey(dbFileKey);
+        String openKeyEntryName = getOpenFileEntryName(volumeId, bucketId, parentID, fileName);
+
+        checkFileState(keyInfo, openKeyEntryName);
+
+        commitKey(dbFileKey, keyInfo, ozoneManager, transactionLogIndex);
+
+        removeOpenKey(openKeyEntryName, transactionLogIndex);
+
+        return openKeyEntryName;
+    }
+
+    private OmKeyInfo getKey(String dbOzoneKey) throws IOException {
+        return omMetadataManager.getKeyTable(getBucketLayout()).get(dbOzoneKey);
+    }
+
+    private String getOpenFileEntryName(long volumeId, long bucketId,
+        long parentObjectId, String fileName) throws IOException {
+        String openFileEntryName = "";
+        String dbOpenKeyPrefix =
+            omMetadataManager.getOpenFileNamePrefix(volumeId, bucketId, parentObjectId, fileName);
+        try (TableIterator<String, ? extends Table.KeyValue<String, OmKeyInfo>>
+            iter = omMetadataManager.getOpenKeyTable(getBucketLayout())
+            .iterator(dbOpenKeyPrefix)) {
+
+            // TODO: verify that there should only be one entry for the open key
```

Comment:
I think multiple file can be uploaded parallel and commit is successful whoever commits first, check this part.

Also cleanup is done for any client, Do any waiting period is required? or cleanup immediately is fine as per HDFS behavior ?

Reply:
This would need to be similar to HDFS